src/pages/index.js

```diff
@@ -54,69 +60,229 @@ const cardSection = new Section(
+  if (isLoading) {
+    buttonElement.textContent = "Saving...";
+  } else {
+    buttonElement.textContent = defaultText;
+  }
+}
+
+//////////////////////////////////// end of project 9 ////////////////////////////////////////////
+
+// //----------------------------------------------------------------------------------------
 //                                     Functions
 //----------------------------------------------------------------------------------------
 function createCard(item) {
-  const card = new Card(item, cardSelector, handleImagePreview);
+  const card = new Card(
+    item,
+    cardSelector,
+    handleImagePreview,
+    (cardId, card) => {
+      deleteCardModal(cardId, card);
+    },
+    (cardId, card) => {
+      handleCardLike(cardId, card);
+    }
+  );
   return card.getview();
 }
 
+// function createCard(item) {
+//   const card = new Card(
+//     item,
+//     cardSelector,
+//     handleImagePreview,
+//     handleCardLike,
+//     (cardId, card) => {
+//       deleteCardModal(cardId, card);
+//     },
+//     (cardId, isLiked, cardElement) => {
+//       handleLikeIcon(cardId, isLiked, cardElement, card);
+//     }
+//   );
+//   return card.getview();
+// }
+
+function handleCardLike(card) {
+  if (!card._isLiked) {
+    api
+      .likeCard(card._id)
+      .then((data) => {
+        console.log(data);
+        card.updateLike(true);
+      })
+      .catch((err) => console.error("Error adding like to card:", err));
+  } else {
+    api
+      .dislikeCard(card._id)
+      .then((data) => {
+        card.updateLike(false);
+      })
+      .catch((err) => console.error("Error removing like from card:", err));
+  }
+}
+
 const editFormValidator = new FormValidator(settings, profileEditForm);
 const addFormValidator = new FormValidator(settings, addCardFormElement);
+const avatarFormValidator = new FormValidator(settings, avatarFormElement);
 
 editFormValidator.enableValidation();
 addFormValidator.enableValidation();
+avatarFormValidator.enableValidation();
 
 function handleImagePreview(cardData) {
   imagePreviewPopup.open(cardData);
 }
+// Project 9
+
+function deleteCardModal(cardId, card) {
+  confirmDelete.setSubmitFunction(() => {
+    api
+      .handleDeleteCard(cardId)
+      .then(() => {
+        card._handleDeleteCard();
+        confirmDelete.close();
+      })
+      .catch((err) => console.error("Error deleting card:", err));
+  });
+
+  confirmDelete.open();
+}
 
 //----------------------------------------------------------------------------------------
 //                                  Event Handlers
 //----------------------------------------------------------------------------------------
 
-function handleProfileEditSubmit(inputData) {
-  // console.log(inputData);
-  // profileTitle.textContent = inputData.name;
-  // profileDescription.textContent = inputData.description;
+function handleProfileEditSubmit(inputData, saveButton) {
+  renderSaving(true, saveButton);
```

Comment:
**COULD BE IMPROVED**

If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}



```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```